### ISSUE TITLE
Abstracted transport layer

### ISSUE THREAD

**Issue by andrewmclagan** [ylletjs/yllet#28]
**Is your feature request related to a problem? Please describe.**
This library is tied to `axios` as the HTTP transport layer. We believe this library can be more flexible if it allows end users to implement their own transport layer - with axios as the default fallback for backwards compatibility.

**Describe the solution you'd like**
A simple abstracted transport layer e.g. (pseudo code)

*Fetch*
```javascript
class FetchTransport {
  /**
   * Transport constructor.
   */
  constructor() {
    // Set up predefined resources methods.
    ["post", "get", "put", "patch", "delete"].forEach(verb => {
      this[verb] = (url, data, config) => this._request(verb, url, data, config);
    });
  }

  _request(verb, url, data, config) {
    const params = {
      method: verb.toUpperCase(),
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify(data)
    };
    const response = fetch(url, params);
    return response.json();
  }
}

```
*Axios*
```javascript
class AxiosTransport {
  /**
   * Transport constructor.
   */
  constructor() {
    this.axios = this._createAxiosClient();
    // Set up predefined resources methods.
    ["post", "get", "put", "patch", "delete"].forEach(verb => {
      this[verb] = (url, data, config) => this.axios[verb](url, data, config);
    });
  }

  /**
   * Create axios client.
   *
   * @return {Axios}
   */
  _createAxiosClient() {
    const auth = Object.assign({}, defaultOptions.auth, this.options.auth);
    const options = deepExtend({}, this.options.axios, {
      auth: auth.username.length || auth.password.length ? auth : null,
      baseURL: this._createBaseUrl(),
      headers: this.options.headers,
    });

    return axios.create(options);
  }  
}
```

**Describe alternatives you've considered**
None

**Additional context**
This is useful as many enduser apps would already be using their own library; superagent, axios, fetch etc.. and loading another library has unnecessary overheads and does not allow interoperability. 

How would we swap them out without loading them? No idea may need a small breaking change to the library so the user `npm install yllet yllet-fetch` or `npm install yllet yllet-axios` then initialises the library as follows

```javascript
import Client from 'yllet';
import AxiosTransport from 'yllet-axios';

const client = new Client({
  endpoint: 'https://demo.wp-api.org/wp-json/',
  transport: AxiosTransport 
});
```

**Comment by frozzare** [comment]
Sounds great, maybe use fetch as a standard so you don't need to install transport layer. This will create a breaking change so it's best to create a new major version with this change.

Feel free to do a pull request.

**Comment by andrewmclagan** [comment]
Awesome. Shall do.

**Comment by andrewmclagan** [comment]
Will start working on this PR, this week.

**Comment by frozzare** [comment]
@andrewmclagan Yes, with this change we can release a new major version (2.0.0).

I think the most important part is that you don't have to install another package to make it work. So it's good to use fetch as the default transport layer 👍

**Comment by andrewmclagan** [comment]
@frozzare Yep - !

Although one thing I would perhaps open a discussion around is the later point; not having to install a second package.

Many many many packages use this approach as there are considerable advantages in such abstractions; A notable example is Babel, they take this approach.

## Explanation

You have to really ask what advantage does not having to install another package offer? Simple one: ease for developers to `yarn add yllet` done. Although they will still have to read the docs to learn how to use the package:

```javascript
import Client from 'yllet';

const client = new Client({
  endpoint: 'https://demo.wp-api.org/wp-json/'
});
```

The alternative approach is a very minor change:

`yarn add @yllet/core @yllet/transport-fetch` 

Then configure:

```javascript
import Client from '@yllet/core';
import FetchTransport from '@yllet/transport-fetch';

const client = new Client({
  endpoint: 'https://demo.wp-api.org/wp-json/',
  transport: new FetchTransport
});
```

The only thing you are loosing is a couple of lines of configuration, developers are used to this, and the alternative approach offers significant advantages:

## Advantages

**Bundle size**

Including a default transport layer such as fetch bloats the package unnecessarily and reduces flexibility.

**Interoperability**

Allowing developers to specify a transport layer they are already using in their app has huge advantages:

Allows them to integrate this client into an existing sdk client they may that accesses other APIs etc without adding a new transport layer... they can use their existing one.

**Testability**

Any changes to the core `@yllet/core` wp-json api client are easily tested as we simple need to mock calls to the transport layer and we dont need to do complex and fragile mocking of HTTP calls.

--------------

So in summary I would argue that this approach actually makes more sense and is inline with wider industry trends and major libraries.

Currently the other major wp-json api client: `wpapi` is a pretty poor offering IMO and has many outstanding issues... Taking a modern modular approach would place this package far ahead of that offering.

Thoughts?

**Comment by andrewmclagan** [comment]
Of course I'm actually already writing other transport layers as I build this PR:

* Axios
* Fetch
* ky

e.g.

```javascript
import Client from '@yllet/core';
import AxiosTransport from '@yllet/transport-axios';

const client = new Client({
  endpoint: 'https://demo.wp-api.org/wp-json/',
  transport: new AxiosTransport
});
```

Its possible to pass along any transport layer specific configs ... easily!

For example https://github.com/sindresorhus/ky has some cool features such as auto retry, simply adding this layer allows an end user to have these features.. pretty cool.

and really ... who knows what other transport layers we could implement: GraphQL etc..

**Comment by andrewmclagan** [comment]
So apologies - although I have started this work today:

Here is an early draft of the PR, follow this fork + .branch to see the progress.

https://github.com/beam-australia/yllet/tree/transports

Basically I will say first off - There are allot of changes here. Not all may make it into the final.

What this includes so far:

## Transport Abstraction

See discussion above.

## Tests refactor

IMO there were some areas where tests were thin on the ground. I have extended and will continue to do so until we reach 95% coverage. 

## Changes in operations / tooling

**YARN** Yeah... so this may seem opinionated. Although simply put `yarn` is just worlds ahead when it comes to monorepo support, there are things NPM simply cant do.

 **JEST** Ava is great, although its main feature - speed through parallelism - is a thing of the past as Jest also does this, has more support for mocking, works with monorepos out of the box and fits into CI tooling allot better.

**@yllet** organisation based packages i the way of the future.

thoughts? 

Big i know - i dont often come in with a wrecking ball but working with monorepos - our team has allot of experience. Personally I'm a core contributor to Material-UI and Apollo both large mono repos. These both suffered through pain of using tooling not designed for this use-case... painful lessons.

**Comment by frozzare** [comment]
I see you point about abstracting `fetch` to a own package, I think it's a valid point and sure we can do it like that. I thought since `fetch` exists in the browser (mostly) we wouldn't have the need of a extra package but when It comes to node we does. 

I don't like the `core` naming part, I think it's better to name the main package `@yllet/yllet` or `@yllet/client`

```js
import Client from '@yllet/yllet';
import Client from '@yllet/client';
```
## Changes in operations / tooling

*Yarn*  - Not sure we need to change this since we use `lerna` and `npm` works very well. 

*Ava* - Sure, we could change to `jest`. I haven't upgrade `ava` since I need to fix the test files and the activity of issues and so has been very low. I have worked with `jest` now also, so I now how that works.

*Yllet organisation* - I like the `@yllet` approach, but I also like monorepon. Monorepos could work and does, I have worked and prefer to work with monorepos and other tools does this aswell, e.g react is one of them. But sure, It could be a pain, but that could repo-based also be if you have a lot of them and the packages should be connected. With monorepos you could automatic bump version since it's the same repo but if the code is in separated repos it's harder.

So I'm not sure about this once, since`@yllet` naming won't match `ylletjs` on github. 

What should the repo naming and structure be if we diseased to go with it?

**Comment by andrewmclagan** [comment]
npm scoping, otherwise known as npm organisations .e.g.

`@yllet/core` or `@yllet/client`... whatever... 

is actually in _support_ of monorepos, I love monorepos. Sorry, I may have not been clear. Not suggesting you move away from monorepos. The opposite. Major libaries have all been migrating to monorepos e.g. Babel, Material-UI, Apollo, React, Vue.... all of them.

npm scoping has no effect on your github repo naming. It would remain the same. `ylletjs/yllet` 

The only difference is on npm where packages are "scoped" to the @yllet organisation. A good example is Babel, in this PR I have actually upgraded babel from 6 to 7. One part of this migration is Babel moving to scoped packages e.g.

```json
{
    "@babel/core": "^7.7.2",
    "@babel/plugin-proposal-class-properties": "^7.7.0",
    "@babel/plugin-proposal-object-rest-spread": "^7.6.2",
    "@babel/plugin-proposal-optional-chaining": "^7.6.0",
    "@babel/preset-env": "^7.7.1",
    "@babel/preset-react": "^7.7.0",    
}
```

Elegant... don't you think?

**Comment by frozzare** [comment]
Yeah I like scoped packages, using that myself. I think I had react as a inspiration when yllet was created and they don't use scoped. But yeah, scoped are a good idea and think we should move to scoped packages.

So that's a great thing to do when creating version 2.0.0 of yllet.

**Comment by andrewmclagan** [comment]
Just for the same of keeping track of the bundle size:

https://bundlephobia.com/result?p=yllet@1.0.2

62% of that is axios.
